core/services/run_queue.go

```diff
@@ -34,81 +35,176 @@ type RunQueue interface {
-	rq.workersMutex.Lock()
-	rq.stopRequested = true
-	rq.workersMutex.Unlock()
-	rq.workersWg.Wait()
+	close(rq.chStop)
+	<-rq.chDone
 }
 
 // Run tells the job runner to start executing a job
 func (rq *runQueue) Run(run *models.JobRun) {
-	rq.workersMutex.Lock()
-	if rq.stopRequested {
-		rq.workersMutex.Unlock()
-		return
+	select {
+	case rq.chRuns <- run:
+	case <-rq.chStop:
 	}
+}
+
+func (rq *runQueue) orchestrateWorkers() {
+	defer close(rq.chDone)
+	for {
+		select {
+		case run := <-rq.chRuns:
+			worker, exists := rq.workers[*run.ID]
+			if !exists {
+				worker = &singleJobSpecWorker{make(chan struct{}), make(chan struct{}), make(chan struct{})}
+				rq.workers[*run.ID] = worker
+				atomic.AddInt32(&rq.numWorkers, 1)
+				go rq.runSingleJobSpecWorker(run.ID, worker)
+				worker.chAddRun <- struct{}{}
+			} else {
+				select {
+				case worker.chAddRun <- struct{}{}:
+				case <-worker.chDone:
+					// If the worker is just spinning down as new work is coming in, allow it
+					// to die and spin up a new one.
+					worker = &singleJobSpecWorker{make(chan struct{}), make(chan struct{}), make(chan struct{})}
+					rq.workers[*run.ID] = worker
+					atomic.AddInt32(&rq.numWorkers, 1)
+					go rq.runSingleJobSpecWorker(run.ID, worker)
+					worker.chAddRun <- struct{}{}
+				}
+			}
+
+		case jobID := <-rq.chWorkerDone:
+			delete(rq.workers, jobID)
+			atomic.AddInt32(&rq.numWorkers, -1)
 
-	runID := run.ID.String()
-	defer numberRunsQueued.Inc()
-	if queueCount, present := rq.workers[runID]; present {
-		rq.workers[runID] = queueCount + 1
-		rq.workersMutex.Unlock()
-		return
+		case <-rq.chStop:
+			for _, run := range rq.workers {
+				run.chStop <- struct{}{}
+				<-run.chDone
+				atomic.AddInt32(&rq.numWorkers, -1)
+			}
+			return
+		}
 	}
-	rq.workers[runID] = 1
-	numberRunQueueWorkers.Set(float64(len(rq.workers)))
-	rq.workersMutex.Unlock()
+}
+
+func (rq *runQueue) runSingleJobSpecWorker(runID *models.ID, worker *singleJobSpecWorker) {
+	defer close(worker.chDone)
 
-	rq.workersWg.Add(1)
+	promNumberRunQueueWorkers.Inc()
+
+	var (
+		startOnce       sync.Once
+		chWorkerStarted = make(chan struct{})
+		chResume        = make(chan int)
+		chWorkComplete  = make(chan struct{})
+	)
+
+	// The worker goroutine accepts job run requests from the coordinator loop below.  It
+	// can accept further requests after work has begun.  Once its queue of work is empty,
+	// it sends a "work finished" message to that loop, which may or may not cause it to
+	// spin down, depending on whether further requests are already enqueued.
 	go func() {
-		for {
-			rq.workersMutex.Lock()
-			queueCount := rq.workers[runID]
-			if queueCount <= 0 {
-				delete(rq.workers, runID)
-				numberRunQueueWorkers.Set(float64(len(rq.workers)))
-				rq.workersMutex.Unlock()
-				break
-			}
-			rq.workers[runID] = queueCount - 1
-			rq.workersMutex.Unlock()
+		for numRunsRequested := range chResume {
+			startOnce.Do(func() { close(chWorkerStarted) })
```

Comment:
If it doesn't change the behavior to move the `Do` before the `for` loop, I think that would be more readable, because it saves readers from wondering why it's being called multiple times.

Reply:
This came up as a result of an issue with how quickly our tests call `AddRun` and then `Stop` (basically, runs get added, but the workers get stopped before they can start any of the work).  To handle this, it's necessary to block the coordinator loop until we have assurance that the worker has actually started doing work.  Wrapping `close(chWorkerStarted)` in a `sync.Once` prevents this from happening more than once.  I agree it's a little awkward, so if you can think of another pattern that word work without creating a race, I'd prefer that.

Reply:
You're right, you have to close it after a value appears in `chResume`. I don't see a cleaner way, off-hand. A comment about why the close happens inside the `for` body might save someone confusion in the future.